plugins/org.locationtech.udig.project.ui/src/org/locationtech/udig/project/ui/internal/tool/display/ToolProxy.java

```diff
@@ -306,78 +305,54 @@ ActionTool getModelessTool() {
-                                        ContributionItem contributionItem = (ContributionItem) optionsContribution;
-                                        contributionItem.setId(contributionId);
-
-                                        addOptionsContribution(
-                                                (ContributionItem) optionsContribution);
-                                    } catch (CoreException e) {
-                                        // TODO Auto-generated catch block
-                                        e.printStackTrace();
-                                    }
-                                }
-
-                            }
+            PlatformGIS.syncInDisplayThread(() -> {
+                try {
+                    Object o = element.createExecutableExtension("class"); //$NON-NLS-1$
+                    tool = (Tool) o;
 
+                    if (tool instanceof AbstractTool) {
+                        ((AbstractTool) tool).init(element);
+                    }
+
+                    /* Tool cursors framework */
+                    if (tool instanceof ModalTool) {
+                        if (defaultCursorID != null) {
+                            ((ModalTool) tool).setCursorID(defaultCursorID);
+                            tool.setProperty(ToolConstants.DEFAULT_CURSOR_ID_KEY,
+                                    defaultCursorID);
                         }
-                    });
-                }
-            };
+                    }
+
+                    tool.setContext(toolContext);
+                    tool.setEnabled(isEnabled());
+
+                    if (tool instanceof ModalTool) {
+                        IMapEditorSelectionProvider selectionProvider = getSelectionProvider();
+                        ModalTool modalTool = (ModalTool) tool;
+                        modalTool.setSelectionProvider(selectionProvider);
+                    }
+
+                    IConfigurationElement[] toolElementChildren = element.getChildren();
+
+                    for (IConfigurationElement toolElement : toolElementChildren) {
+
+                        if (toolElement.getAttribute("class") != null) {
+                            String contributionId = toolElement.getAttribute("id");
 
-            PlatformGIS.syncInDisplayThread(runnable);
+                            Object optionsContribution = toolElement
+                                    .createExecutableExtension("class");
+                            ContributionItem contributionItem = (ContributionItem) optionsContribution;
+                            contributionItem.setId(contributionId);
+
+                            addOptionsContribution((ContributionItem) optionsContribution);
+                        }
+                    }
+
+                } catch (CoreException e) {
+                    tool = null;
+                    ProjectUIPlugin.log("Error loading tool", e); //$NON-NLS-1$
```

Comment:
Way better to log it, this allows users easier to find out which Tool-Extension causing trouble. Any chance to get the extension-id (which is a required attribute for modalTool, actionTool, and backgroundTool)?

Reply:
done